common/src/main/java/io/netty/util/DomainWildcardMappingBuilder.java

```diff
@@ -0,0 +1,135 @@
+     *
+     * @param initialCapacity initial capacity for the internal map
+     * @param defaultValue    the default value for {@link Mapping#map(Object)} to return
+     *                        when nothing matches the input
+     */
+    public DomainWildcardMappingBuilder(int initialCapacity, V defaultValue) {
+        this.defaultValue = checkNotNull(defaultValue, "defaultValue");
+        map = new LinkedHashMap<String, V>(initialCapacity);
+    }
+
+    /**
+     * Adds a mapping that maps the specified (optionally wildcard) host name to the specified output value.
+     * {@code null} values are forbidden for both hostnames and values.
+     * <p>
+     * <a href="https://tools.ietf.org/search/rfc6125#section-6.4">DNS wildcard</a> is supported as hostname. The
+     * wildcard will only match one sub-domain deep and only when wildcard is used as the most-left label.
+     *
+     * For example:
+     *
+     * <p>
+     *  *.netty.io will match xyz.netty.io but NOT abc.xyz.netty.io
+     * </p>
+     *
+     * @param hostname the host name (optionally wildcard)
+     * @param output   the output value that will be returned by {@link Mapping#map(Object)}
+     *                 when the specified host name matches the specified input host name
+     */
+    public DomainWildcardMappingBuilder<V> add(String hostname, V output) {
+        map.put(ImmutableDomainWildcardMapping.normalize(checkNotNull(hostname, "hostname")),
+                checkNotNull(output, "output"));
+        return this;
+    }
+
+    /**
+     * Creates a new instance of an immutable {@link Mapping}.
+     *
+     * @return new {@link Mapping} instance
+     */
+    public Mapping<String, V> build() {
+        return new ImmutableDomainWildcardMapping<V>(defaultValue, new LinkedHashMap<String, V>(map));
+    }
+
+    private static final class ImmutableDomainWildcardMapping<V> implements Mapping<String, V> {
+        private static final String REPR_HEADER = "ImmutableDomainWildcardMapping(default: ";
+        private static final String REPR_MAP_OPENING = ", map: ";
+        private static final String REPR_MAP_CLOSING = ")";
+
+        private final V defaultValue;
+        private final Map<String, V> map;
+
+        ImmutableDomainWildcardMapping(V defaultValue, Map<String, V> map) {
+            this.defaultValue = defaultValue;
+            this.map = Collections.unmodifiableMap(map);
+        }
+
+        @Override
+        public V map(String hostname) {
+            if (hostname != null) {
+                hostname = normalize(hostname);
+
+                // Let's try an exact match first
+                V value = map.get(hostname);
+                if (value != null) {
+                    return value;
+                }
+
+                // No exact match, let's try a wildcard match.
+                int idx = hostname.indexOf('.');
+                if (idx != -1) {
+                    value = map.get('*' + hostname.substring(idx));
```

Comment:
Optional idea:  If a user adds a wildcard domain `*.netty.io`,  it could be normalized to `.netty.io`, and stored in the map.  That would make it cheaper for this line to check if it's present, because no string concatenation would be needed.  You could just check the substring.

I am not sure how fast you want it to go.

Reply:
thats a good idea ... Let me do this